### PR TITLE
[GEOT-6510] Upgrade Apache Commons Collections from v3 to v4

### PR DESCRIPTION
**Pull request by soc**
This change migrates code inside Geotools itself, as well as updating
Apache Commons Digester to the first newer version that doesn't depend
on Apache Commons Collections v3 directly anymore.

(Apache Commons Digester still depends on Apache Commons Collections v3
via Apache Commons BeanUtils, but this will be fixable in the future
after Apache Commons BeanUtils v2 (which depends on Apache Commons
Collections v4) is released.

### CODE REVIEW

**Comment by soc** [comment]
- [x] Confirm you have read the [contribution guidelines](https://github.com/geotools/geotools/blob/master/CONTRIBUTING.md) 
- [x] ~You have sent a Contribution Licence Agreement (CLA) as necessary~ (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.
- [ ] The changes are not breaking the build in downstream projects using SNAPSHOT dependencies, GeoWebCache and GeoServer.

- [x] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[GEOT-XYZW] Title of the Jira ticket"
- [ ] New unit tests have been added covering the changes
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html) (QA checks results will be reported by travis-ci after opening this PR)
- [x] ~Documentation has been updated accordingly.~ (not applicable)

**Comment by soc** [comment]
I have no idea what's happening here:

```
[ERROR] Tests run: 3, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 2.159 s <<< FAILURE! - in org.geotools.kml.KMLParsingTest
[ERROR] testEncodeFeatureCollection(org.geotools.kml.KMLParsingTest)  Time elapsed: 0.061 s  <<< ERROR!
java.lang.ClassCastException: org.geotools.feature.collection.BridgeIterator cannot be cast to java.util.Collection
	at org.geotools.kml.KMLParsingTest.testEncodeFeatureCollection(KMLParsingTest.java:99)
[ERROR] Tests run: 4, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 0.018 s <<< FAILURE! - in org.geotools.kml.v22.KMLEncodingTest
[ERROR] testEncodeSimpleFeaturecollection(org.geotools.kml.v22.KMLEncodingTest)  Time elapsed: 0.004 s  <<< ERROR!
java.lang.ClassCastException: org.geotools.feature.collection.BridgeIterator cannot be cast to java.util.Collection
	at org.geotools.kml.v22.KMLEncodingTest.testEncodeSimpleFeaturecollection(KMLEncodingTest.java:67)

[ERROR] testEncodeNullAttribute(org.geotools.kml.v22.KMLEncodingTest)  Time elapsed: 0.003 s  <<< ERROR!
java.lang.ClassCastException: org.geotools.feature.collection.BridgeIterator cannot be cast to java.util.Collection
	at org.geotools.kml.v22.KMLEncodingTest.testEncodeNullAttribute(KMLEncodingTest.java:129)
```

my changes are completely unrelated and they didn't show after the first commit (the updated commit was only to remove a deprecation warning).

**Comment by soc** [comment]
Will have to investigate this test failure:

```
[ERROR] Tests run: 3, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 2.171 s <<< FAILURE! - in org.geotools.kml.KMLParsingTest
[ERROR] testEncodeFeatureCollection(org.geotools.kml.KMLParsingTest)  Time elapsed: 0.062 s  <<< ERROR!
java.lang.ClassCastException: org.geotools.feature.collection.BridgeIterator cannot be cast to java.util.Collection
	at org.geotools.kml.KMLParsingTest.testEncodeFeatureCollection(KMLParsingTest.java:99)
```

**Comment by aaime** [comment]
Had a one second check, there is lots of white space changes, please back them out before asking for a review again, thanks!

**Comment by soc** [comment]
@aaime Done!

**Comment by aaime** [comment]
The GeoTools, GeoWebCache and GeoServer projects act as a unit, they are linked via snapshot dependencies (so they live depend on each other). This is due to GT/GWC getting the vast majority of the changes out of GeoServer activity.

I checked downstreams, GeoServer uses an older version of digester (same as current GeoTools), wondering, was the upgrade necessary and related to commons-collections upgrade?

Besides that, better to wait until the 23-RC release is done, this upgrade may require changes in both downstream projects... though I guess both versions can co-exist?
I see that GWC uses CaseInsensitiveMap, while GeoServer usage seems more extensive:

```
geoserver> git grep "commons.collections" | grep java

community/jdbcconfig/src/main/java/org/geoserver/jdbcconfig/internal/ConfigDatabase.java:import org.apache.commons.collections.CollectionUtils;
community/jdbcconfig/src/main/java/org/geoserver/jdbcconfig/internal/ConfigDatabase.java:import org.apache.commons.collections.Predicate;
community/notification-geonode/src/main/java/org/geoserver/notification/geonode/GeoNodeJsonEncoder.java:import org.apache.commons.collections.CollectionUtils;
community/notification/src/main/java/org/geoserver/notification/NotificationCatalogListener.java:import org.apache.commons.collections.CollectionUtils;
community/oseo/oseo-core/src/main/java/org/geoserver/opensearch/eo/OSEOFilter.java:import org.apache.commons.collections.map.CaseInsensitiveMap;
community/qos/src/main/java/org/geoserver/qos/BaseQosXmlEncoder.java:import org.apache.commons.collections4.CollectionUtils;
community/qos/src/main/java/org/geoserver/qos/QosMainConfigurationWMSValidator.java:import org.apache.commons.collections4.CollectionUtils;
community/qos/src/main/java/org/geoserver/qos/web/BaseLimitedConstraintsPanel.java:import org.apache.commons.collections4.CollectionUtils;
community/qos/src/main/java/org/geoserver/qos/wfs/WfsQosXmlEncoder.java:import org.apache.commons.collections4.CollectionUtils;
community/qos/src/main/java/org/geoserver/qos/wms/WmsQosXmlEncoder.java:import org.apache.commons.collections4.CollectionUtils;
community/solr/src/main/java/org/geoserver/solr/SolrAttributeProvider.java:import org.apache.commons.collections.CollectionUtils;
extension/app-schema/app-schema-mongo-test/src/test/java/org/geoserver/test/onlineTest/ComplexMongoDBTest.java:import org.apache.commons.collections.CollectionUtils;
extension/importer/bdb/src/main/java/org/geoserver/importer/bdb/BDBImportStore.java:import org.apache.commons.collections.Predicate;
extension/importer/bdb/src/main/java/org/geoserver/importer/bdb/BDBImportStore.java:import org.apache.commons.collections.iterators.FilterIterator;
extension/importer/web/src/main/java/org/geoserver/importer/web/ImportContextProvider.java:import org.apache.commons.collections.IteratorUtils;
extension/wps/wps-cluster-hazelcast/src/main/java/org/geoserver/wps/hz/FilterToCriteria.java:import org.apache.commons.collections.functors.FalsePredicate;
main/src/main/java/org/geoserver/config/util/XStreamPersister.java:import org.apache.commons.collections.map.MultiValueMap;
main/src/main/java/org/geoserver/security/DisabledServiceResourceFilter.java:import org.apache.commons.collections.CollectionUtils;
restconfig/src/main/java/org/geoserver/rest/FontListController.java:import org.apache.commons.collections.map.HashedMap;
wcs/src/main/java/org/geoserver/wcs/responses/BaseCoverageResponseDelegate.java:import org.apache.commons.collections.map.CaseInsensitiveMap;
wcs2_0/src/test/java/org/geoserver/wcs2_0/kvp/WCSKVPTestSupport.java:import org.apache.commons.collections.map.CaseInsensitiveMap;
web/core/src/main/java/org/geoserver/web/data/resource/FeatureResourceConfigurationPanel.java:import org.apache.commons.collections.CollectionUtils;
web/demo/src/main/java/org/geoserver/web/demo/MapPreviewPage.java:import org.apache.commons.collections4.CollectionUtils;
wfs/src/main/java/org/geoserver/wfs/xml/GML2OutputFormat2.java:import org.apache.commons.collections.map.MultiValueMap;
wms/src/main/java/org/geoserver/wms/legendgraphic/ColorMapLegendCreator.java:import org.apache.commons.collections.map.CaseInsensitiveMap;
wms/src/main/java/org/geoserver/wms/map/GetMapKvpRequestReader.java:import org.apache.commons.collections.EnumerationUtils;
```

**Comment by soc** [comment]
> I checked downstreams, GeoServer uses an older version of digester (same as current GeoTools), wondering, was the upgrade necessary and related to commons-collections upgrade?

I think it could be left out for this change – it was to update to the last available 1.x release (2005 → 2009).

> Besides that, better to wait until the 23-RC release is done, this upgrade may require changes in both downstream projects... though I guess both versions can co-exist?

Yes, the versions can coexist (and do already) – as seen in the results of your grep both versions are used already

> I see that GWC uses CaseInsensitiveMap, while GeoServer usage seems more extensive:

From a first look, there doesn't seem to be anything in there that can't be upgraded by replacing `org.apache.commons.collections.` with `org.apache.commons.collections4.`.

**Comment by jodygarnett** [comment]
This release candidate goes out soon, I am keen to see this work go ahead after :)

**Comment by soc** [comment]
@jodygarnett Great, thanks!

**Comment by aaime** [comment]
I've checked GeoServer a bit better, it depends on commons-collections4 too, through Wicket, but on version 4.1. There is no dependency management for it. @soc do you think this change in GeoTools would work with 4.1 too, to keep versions aligned? In alternative, can you check the commons-collection 4.x history and verify there is no breaking change that would cause Wicket not to work any more with 4.4?

**Comment by soc** [comment]
@aaime 

> do you think this change in GeoTools would work with 4.1 too, to keep versions aligned

Yes, this should work. (Though I prefer your latter suggestion:)

> can you check the commons-collection 4.x history and verify there is no breaking change that would cause Wicket not to work any more with 4.4

Done – there are no incompatibilities between 4.4 and 4.1.

**Comment by jodygarnett** [comment]
Nice work, any additional feedback Andrea? Do we have anything downstream to fix ...

File: modules/extension/xsd/xsd-core/src/main/java/org/geotools/xsd/impl/BindingPropertyExtractor.java [hunk 1]
```diff
@@ -116,28 +115,28 @@ public List properties(Object object, XSDElementDeclaration element) {
                 // turn each map entry into a particle
                 HashMap particles = new HashMap();
 
-                for (Iterator e = map.entrySet().iterator(); e.hasNext(); ) {
-                    Map.Entry entry = (Map.Entry) e.next();
+                for (Iterator e = map.keySet().iterator(); e.hasNext(); ) {
```
Comment (jodygarnett) [inline modules/extension/xsd/xsd-core/src/main/java/org/geotools/xsd/impl/BindingPropertyExtractor.java:118]: why change to a keyset if you are going to do a map.get(key) later anyways?

Comment (soc) [reply modules/extension/xsd/xsd-core/src/main/java/org/geotools/xsd/impl/BindingPropertyExtractor.java:118]: The new Map type doesn't have an `entrySet()` method, and this was the least ugly way to do it. `mapIterator` for example has different semantics. I'm not too sure on the details anymore, but I remember spending some time pulling my hair out on this change.

**Comment by aaime** [comment]
@jodygarnett yep, I think the missing bit is the downstream migration (GS and GWC). That said, the build is green, as already noted before, the two libraries can coexist.

**Comment by aaime** [comment]
GWC upgrade appears to be trivial: https://github.com/GeoWebCache/geowebcache/pull/849

**Comment by aaime** [comment]
Possibly incomplete GeoServer upgrade PR here: https://github.com/geoserver/geoserver/pull/4232

File: modules/extension/xsd/xsd-core/src/main/java/org/geotools/xsd/impl/BindingPropertyExtractor.java (hunk 1 above)
Comment (aaime) [reply modules/extension/xsd/xsd-core/src/main/java/org/geotools/xsd/impl/BindingPropertyExtractor.java:118]: As an alternative, the MultiValuedMap can return a MapIterator, which returns the key when calling "next", and has a dedicated "getValue" method to fetch the value associated to the current entry.

**Comment by aaime** [comment]
@soc this PR changes several files, a CLA would be required for merging. @jodygarnett confirm?

**Comment by soc** [comment]
@aaime Not sure if these changes reach the threshold of originality; if you need a CLA – my org has already submitted one a few years ago, but my name was not listed in schedule A back then. What's the process here?

**Comment by aaime** [comment]
Can you tell us what organization is that? If you're working as an employee there and this is done in working hours, your org CLA should be covering your activity already, right @jodygarnett ?

**Comment by aaime** [comment]
@soc any feedback?

**Comment by soc** [comment]
@aaime I sent you an email to the address listed in your GitHub profile. Thanks!

**Pull request merged**